### DISCUSSION TITLE
Incremental queries

### DISCUSSION THREAD

**Discussion by oxalica** [tree-sitter/tree-sitter#1976] [Ideas] [answered]
I found that `Query` doesn't have caches or states to reuse information between edits, and always traverses the full tree and takes a lot of time in [`ts_query_cursor__advance`](https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/query.c#L3221). (Also there's [a bug](https://github.com/tree-sitter/tree-sitter/issues/1972) making it slower)
This is acceptable for one-pass queries like in `tree-sitter-highlight` CLI. But it causes performance issues more and more often when tree-sitter are integrated in editors like [`helix`](https://github.com/helix-editor/helix) and [`nvim-treesitter`](https://github.com/nvim-treesitter/nvim-treesitter).

Note that `{start,end}_{byte,point}` doesn't help here. It can only limit the result size rather than the _search space_, let alone this limit is only respected [under some conditions](https://github.com/tree-sitter/tree-sitter/blob/870fb8772f3e47b5aef4c7000a78d61f3aa2b005/lib/src/query.c#L3355-L3361). For a "non-local" query like `(foo . bar  . qux @hl)` or `(foo (bar) (qux) @hl)`, if `foo` are at the beginning of the file (or spans almost the full file) while `qux` is at the end and in view-port, we have to check everything between to determine if the query matches `bar`, although they are not visible. 

I'd like to ask for incremental queries, which maintains the result of queries and supports `edit` in less than `O(full_tree_size)` time, just like `edit`/`parse` for `TSTree`. So we could get "the diff of query results" and patch the highlighting quickly. This seems computationally very hard and I'm not sure if this time-complexity goal is achievable. So I post here for discussion.

**Comment by theHamsta** [comment]
I would have also expected that you would need to specify a node further nested in the tree to actually reduce the query overhead. I believe in Neovim we only limit the line range

**Reply by oxalica** [reply]
> I would have also expected that you would need to specify a node further nested in the tree to actually reduce the query overhead.

Many huge nodes still intersects the given range, so it doesn't help. Here's [an example](https://raw.githubusercontent.com/NixOS/nixpkgs/04e7308c6833a1f0a58dcdbd12309c7859f32d32/pkgs/top-level/all-packages.nix),
```nix
pkgs:
/* ... */
{ # <- This (attrset_expression) node covers almost the full file.
   pkg1 = /*...*/;
   pkgs2 = /*...*/;
   # 38,000 lines go here.
   pkgsn = /*...*/; # <- Edit here.
}
```

Then there's a query which scans because `attrset_expression` intersects the range. Note that we cannot remove the parent `attrset_expression` node since `binding` appeared in other nodes shouldn't be highlighted as `@field`.
```
(attrset_expression ;; <- This covers almost the full file.
 (binding_set       ;; <- So does this.
  (binding . (attrpath (identifier) @field)))) ;; <- This is limited by range.
```

Matching is currently done top-down. But we somehow need to do it bottom-up to limit the node to be traversed.

**Comment by maxbrunsfeld** [comment]
Setting a start byte and end byte (or start and end points) on the query cursor *does* limit the set of nodes traversed. The cursor only visits nodes if they intersect the given range, or if a partial match is currently in progress, and was started on a node that *did* intersect the range.

**Reply by oxalica** [reply]
> if they intersect the given range

Unfortunately, many nodes are so big that **always** intersect the given range.
You can discover that with a simple test.

```rust
use anyhow::Result;
use libloading::Library;
use std::time::Instant;
use tree_sitter::{Language, Parser, Point, Query, QueryCursor};

fn main() -> Result<()> {
    // https://github.com/cstrahan/tree-sitter-nix
    let lib = unsafe { Library::new("/path/to/your/parser.so")? };
    let lang = unsafe { lib.get::<extern "C" fn() -> Language>(b"tree_sitter_nix")?() };
    let mut parser = Parser::new();
    parser.set_language(lang)?;

    // curl 'https://raw.githubusercontent.com/NixOS/nixpkgs/04e7308c6833a1f0a58dcdbd12309c7859f32d32/pkgs/top-level/all-packages.nix'
    let text = std::fs::read_to_string("all-packages.nix")?;
    let tree = parser.parse(&text, None).unwrap();

    let query = r"
(attrset_expression ;; <- This covers almost the full file.
 (binding_set       ;; <- So does this.
  (binding . (attrpath (identifier) @field)))) ;; <- This is limited by range.
    ";
    let query = Query::new(lang, query)?;

    let time = Instant::now();
    let mut matched = Vec::new();
    let mut cursor = QueryCursor::new();
    // Set range. The selected content are,
    // ```
    // 38111
    // 38112 honeytail = callPackage ../servers/tracing/honeycomb/honeytail { };
    // 38113
    // ```
    cursor.set_point_range(Point::new(38110, 0)..Point::new(38112, 0));
    for m in cursor.matches(&query, tree.root_node(), text.as_bytes()) {
        for cap in m.captures {
            matched.push((cap.node.start_position(), cap.node.end_position()));
        }
    }
    println!("{:?} {}", time.elapsed(), matched.len());
    Ok(())
}
```

Run with `cargo run --release` and the result is,
```
16.809816ms 15494
```
I'm not convinced there are 15494 matches in that three lines, where 2 lines are actually empty.
If you delete the `set_point_range` line, the the result is,
```
35.854629ms 18568
```
Ok, that's even slower. You can see here that the range limit rules out very little nodes. Most of the nodes is still scanned and returned.

The 16ms cost is not that slow, but notice that I only use one query while tens or hundreds queries are used in production, let alone queries are done in each frame for highlighting.

**Reply by maxbrunsfeld** [reply, marked as answer]
I think this is improved by https://github.com/tree-sitter/tree-sitter/pull/2085

**Comment by vimpostor** [comment]
I think this is fixed with [#22309](https://github.com/neovim/neovim/pull/22309)
